File: Coiffeur/src/Document.m
//
//  Document.m
//  Coiffeur
//
//  Created by Anton Leuski on 3/25/15.
//  Copyright (c) 2015 Anton Leuski. All rights reserved.
//

#import "Document.h"

#import "ALMainWindowController.h"
#import "ALCoiffeurController.h"
#import "ALCoiffeurView.h"

#pragma clang diagnostic push
#pragma ide diagnostic ignored "OCNotLocalizedStringInspection"
static NSString* const ALCoiffeurErrorDomain = @"Coiffeur";
#pragma clang diagnostic pop

@interface Document ()
@property (nonatomic, strong) ALCoiffeurView* coiffeur;
@end

@implementation Document

- (instancetype)initWithType:(NSString*)typeName error:(NSError**)outError
{
  if (self = [super initWithType:typeName error:outError]) {
    self.model = [self AL_modelControllerOfType:typeName error:outError];

    if (!self.model) {
      self = nil;
    }
  }

  return self;
}

- (NSUndoManager*)undoManager
{
  return self.model ? self.model.managedObjectContext.undoManager : [super undoManager];
}

- (ALCoiffeurController*)AL_modelControllerOfType:(NSString*)type error:(NSError**)outError
{
  for (Class c in[ALCoiffeurController availableTypes]) {
    if ([type isEqualToString:[c documentType]]) {
      return [[c alloc] initWithError:outError];
    }
  }

  if (outError) {
    NSString* description
      = [NSString stringWithFormat:NSLocalizedString(@"Unknown document type “%@”", NULL), type];
    *outError = [NSError errorWithDomain:ALCoiffeurErrorDomain
                                    code:0
                                userInfo:@{NSLocalizedDescriptionKey: description}];
  }

  return nil;
}

- (BOOL)AL_ensureWeHaveModelOfType:(NSString*)typeName
                    errorFormatKey:(NSString*)errorFormatKey
                             error:(NSError**)outError
{
  if (self.model) {
    if (![typeName isEqualToString:self.model.documentType]) {
      if (outError) {
        NSString* description = [NSString stringWithFormat:NSLocalizedString(errorFormatKey, NULL)
                                 , typeName, self.model.documentType];
        *outError = [NSError errorWithDomain:ALCoiffeurErrorDomain
                                        code:0
                                    userInfo:@{NSLocalizedDescriptionKey: description}];
      }

      return NO;
    }
  } else {
    self.model = [self AL_modelControllerOfType:typeName error:outError];

    if (!self.model) {
      return NO;
    }
  }

  return YES;
}

- (BOOL)readFromURL:(NSURL*)url ofType:(NSString*)typeName error:(NSError**)outError
{
  if (![self AL_ensureWeHaveModelOfType:typeName
                         errorFormatKey:@"Cannot read content of document “%@” into document “%@”"
                                  error:outError])
  {
    return NO;
  }

  return [self.model readValuesFromURL:url error:outError];
}

- (BOOL)writeToURL:(NSURL*)url ofType:(NSString*)typeName error:(NSError**)outError
{
  if (![self AL_ensureWeHaveModelOfType:typeName
                         errorFormatKey:@"Cannot write content of document “%2$@” as “%1$@”"
                                  error:outError])
  {
    return NO;
  }

  return [self.model writeValuesToURL:url error:outError];
}

+ (BOOL)autosavesInPlace
{
  return YES;
}

- (void)makeWindowControllers
{
  ALMainWindowController* controller = [ALMainWindowController new];

  [self addWindowController:controller];
}

- (void)embedInView:(NSView*)container
{
  self.coiffeur = [[ALCoiffeurView alloc] initWithModel:self.model bundle:nil];
  [self.coiffeur embedInView:container];
}

- (void)canCloseDocumentWithDelegate:(id)delegate
                 shouldCloseSelector:(SEL)shouldCloseSelector
                         contextInfo:(void*)contextInfo
{
  [self.model.managedObjectContext commitEditing];
  [super canCloseDocumentWithDelegate:delegate
                  shouldCloseSelector:shouldCloseSelector
                          contextInfo:contextInfo];
}

- (NSArray*)writableTypesForSaveOperation:(NSSaveOperationType)saveOperation
{
  return @[self.model.documentType];
}

@end

File: Coiffeur/src/Document.swift
//
//  Document.swift
//  Coiffeur
//
//  Created by Anton Leuski on 4/7/15.
//  Copyright (c) 2015 Anton Leuski. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//
//

import Cocoa

@objc(Document)
class Document: NSDocument {

  var model: CoiffeurController?

  override init()
  {
    super.init()
  }

  convenience init(type typeName: String) throws
  {
    self.init()
    self.fileType = typeName
    self.model = try CoiffeurController.coiffeurWithType(typeName)
  }

  override var undoManager: UndoManager? {
    get {
      if let model = self.model {
        return model.managedObjectContext.undoManager
      } else {
        return super.undoManager
      }
    }
    set {
      super.undoManager = newValue
    }
  }

  private func _ensureWeHaveModelOfType(
    _ typeName: String) throws -> String
  {
    if let model = self.model {
      return type(of: model).documentType
    }

    self.model = try CoiffeurController.coiffeurWithType(typeName)
    return typeName
  }

  override func read(from absoluteURL: URL, ofType typeName: String) throws
  {
    let modelTypeName = try _ensureWeHaveModelOfType(typeName)

    if modelTypeName != typeName {
      throw Errors.cannotReadAs(typeName, modelTypeName)
    }

    try self.model?.readValuesFromURL(absoluteURL)
  }

  override func write(to absoluteURL: URL, ofType typeName: String) throws
  {
    let modelTypeName = try _ensureWeHaveModelOfType(typeName)

    if modelTypeName != typeName {
      throw Errors.cannotWriteAs(typeName, modelTypeName)
    }

    try self.model?.writeValuesToURL(absoluteURL)
  }

  override class var autosavesInPlace: Bool
  {
    return true
  }

  override func makeWindowControllers()
  {
    self.addWindowController(MainWindowController())
  }

  //  override func canCloseDocumentWithDelegate(delegate: AnyObject,
  //    shouldCloseSelector: Selector, contextInfo: UnsafeMutablePointer<Void>)
  //  {
  //    self.model?.managedObjectContext.commitEditing()
  //    super.canCloseDocumentWithDelegate(delegate,
  //      shouldCloseSelector:shouldCloseSelector, contextInfo:contextInfo)
  //  }

  override func writableTypes(for _: NSDocument.SaveOperationType)
    -> [String]
  {
    if let model = self.model {
      return [type(of: model).documentType]
    } else {
      return []
    }
  }

}
